plugins/chain_plugin/chain_plugin.cpp

```diff
@@ -779,9 +789,24 @@ void chain_plugin::plugin_initialize(const variables_map& options) {
          auto infile = std::ifstream(my->snapshot_path->generic_string(), (std::ios::in | std::ios::binary));
          auto reader = std::make_shared<istream_snapshot_reader>(infile);
          reader->validate();
-         reader->read_section<genesis_state>([this]( auto &section ){
-            section.read_row(my->chain_config->genesis);
+         chain_snapshot_header header;
```

Comment:
Create a helper function (perhaps in controller) that takes a snapshot reader and returns an extracted chain ID. The snapshot related code below would (with some modifications) go into that new function.